routers/api/v2/users.go

```diff
@@ -280,6 +281,97 @@ func (r *apiV2Router) RemoveFromTeam(ctx *gin.Context) {
 	ctx.Status(http.StatusNoContent)
 }
 
+// PUT: /api/v2/users/(:id|me)/password
+// x-www-form-urlencoded
+// Request:  password string
+// Response:
+// Headers:  Authorization -> token
+func (r *apiV2Router) SetPassword(ctx *gin.Context) {
+	var req struct {
+		Password string `form:"password"`
+	}
+	_ = ctx.Bind(&req)
+	if len(req.Password) == 0 {
+		r.logger.Debug("password not specified", zap.Int("password length", len(req.Password)))
+		models.SendAPIError(ctx, http.StatusBadRequest, "request must include the new password")
+		return
+	}
+
+	user, err := r.getUserCtxAware(ctx, ctx.Param("id"))
+	if err != nil {
+		switch errors.Cause(err) {
+		case v2.ErrInvalidToken:
+			r.logger.Debug("invalid token", zap.Error(err))
+			r.HandleUnauthorized(ctx)
+		case v2.ErrInvalidTokenType:
+			r.logger.Debug("invalid token type", zap.Error(err))
+			models.SendAPIError(ctx, http.StatusBadRequest, "provided token is of invalid type for the requested operation")
+		case services.ErrInvalidID:
+			r.logger.Debug("invalid user id", zap.Error(err))
+			models.SendAPIError(ctx, http.StatusBadRequest, "invalid user id")
+		case services.ErrNotFound:
+			r.logger.Debug("user not found", zap.Error(err))
+			models.SendAPIError(ctx, http.StatusNotFound, "user not found")
+		default:
+			r.logger.Error("could not fetch user", zap.Error(err))
+			models.SendAPIError(ctx, http.StatusInternalServerError, "something went wrong")
+		}
+		return
+	}
+
+	err = r.userService.ResetPasswordForUserWithIDAndEmail(ctx, user.ID.Hex(), user.Email, req.Password)
```

Comment:
It would be great if we replaced this with a call to `UserService.UpdateUserWithID`. We'd be able to get away with a single call to `UserService` since we wouldn't need to fetch the email address.

I'd want each operation to make at most 1 call to any given service. The idea here is that if we update the MongoDB service clients to be transactional, all operations would automatically be transactional as well. We can't really do transactions at the router level because transactions would be technology dependent whereas the router should be technology agnostic so we have to implement transactional logic at the service client level.

Reply:
The `SetPassword` method would still need to make two calls to the user service, in the case that `id=me` is provided we'd need to get the user based on the token.

I can change the call to `UpdateUserWithID` but just to note that there would still be two calls to the user service in some cases.

Reply:
You can call `GetUserIdFromToken` on the authorizer to extract the user id from the token without a  call to external services and then you can just use that id in the call to `UpdateUserWithID`

Reply:
Ohhh, I see! That makes sense 👍